From f0a6d2dfd242fa1e5ce7c72523f7c532e037a147 Mon Sep 17 00:00:00 2001
From: Taozuhong <taozuhong@hotmail.com>
Date: Sun, 11 Oct 2026 09:24:33 +0000
Subject: [PATCH 2/2] libpas/MinGW: explicit windows.h includes and PE-friendly
 symbol visibility

- pas_committed_pages_vector.c, pas_page_malloc.c, pas_scavenger.c and
  pas_probabilistic_guard_malloc_allocator.c use WinAPI under
  PAS_OS(WINDOWS) (VirtualQuery, VirtualAlloc/VirtualFree,
  QueryPerformance*) without including <windows.h>; Apple's Windows
  build obtains it from a prefix header, MinGW builds do not, so
  include it explicitly.
- pas_utils_prefix.h: __PAS_API expanded to visibility attributes.
  PE/COFF has no visibility semantics and MinGW GCC emits
  'visibility attribute not supported in this configuration; ignored'
  for every annotated declaration, flooding the build. libpas is linked
  as a static archive, so drop the attribute entirely on MinGW.
---
 .../libpas/src/libpas/pas_committed_pages_vector.c        | 4 ++++
 Source/bmalloc/libpas/src/libpas/pas_page_malloc.c        | 4 ++++
 .../src/libpas/pas_probabilistic_guard_malloc_allocator.c | 2 ++
 Source/bmalloc/libpas/src/libpas/pas_scavenger.c          | 4 ++++
 Source/bmalloc/libpas/src/libpas/pas_utils_prefix.h       | 8 +++++++-
 5 files changed, 21 insertions(+), 1 deletion(-)

diff --git a/Source/bmalloc/libpas/src/libpas/pas_committed_pages_vector.c b/Source/bmalloc/libpas/src/libpas/pas_committed_pages_vector.c
index f2e5d1ca..3191aa5b 100644
--- a/Source/bmalloc/libpas/src/libpas/pas_committed_pages_vector.c
+++ b/Source/bmalloc/libpas/src/libpas/pas_committed_pages_vector.c
@@ -36,6 +36,10 @@
 #include <sys/mman.h>
 #endif
 
+#if PAS_OS(WINDOWS)
+#include <windows.h> /* for MEMORY_BASIC_INFORMATION / VirtualQuery */
+#endif
+
 PAS_BEGIN_EXTERN_C;
 
 void pas_committed_pages_vector_construct(pas_committed_pages_vector* vector,
diff --git a/Source/bmalloc/libpas/src/libpas/pas_page_malloc.c b/Source/bmalloc/libpas/src/libpas/pas_page_malloc.c
index ac0e48a4..99334f9c 100644
--- a/Source/bmalloc/libpas/src/libpas/pas_page_malloc.c
+++ b/Source/bmalloc/libpas/src/libpas/pas_page_malloc.c
@@ -36,6 +36,10 @@
 #include <sys/mman.h>
 #include <unistd.h>
 #endif
+
+#if PAS_OS(WINDOWS)
+#include <windows.h> /* for VirtualAlloc / VirtualFree / VirtualQuery / GetSystemInfo */
+#endif
 #if PAS_OS(DARWIN)
 #include <mach/vm_statistics.h>
 #endif
diff --git a/Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c b/Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c
index fb8709a6..378c33c9 100644
--- a/Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c
+++ b/Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c
@@ -40,6 +40,8 @@
 #include <stdlib.h>
 #if !PAS_OS(WINDOWS)
 #include <sys/mman.h>
+#else
+#include <windows.h> /* for VirtualAlloc / VirtualFree / VirtualQuery */
 #endif
 
 /* PlayStation does not currently support the backtrace API. Android API versions < 33 don't, either. Windows does not either. Linux only with GLibc and not uCLibc/Musl. */
diff --git a/Source/bmalloc/libpas/src/libpas/pas_scavenger.c b/Source/bmalloc/libpas/src/libpas/pas_scavenger.c
index ee24c307..eb62547f 100644
--- a/Source/bmalloc/libpas/src/libpas/pas_scavenger.c
+++ b/Source/bmalloc/libpas/src/libpas/pas_scavenger.c
@@ -29,6 +29,10 @@
 
 #include "pas_scavenger.h"
 
+#if PAS_OS(WINDOWS)
+#include <windows.h> /* for QueryPerformanceFrequency / QueryPerformanceCounter */
+#endif
+
 #include <math.h>
 #include "pas_baseline_allocator_table.h"
 #include "pas_compact_expendable_memory.h"
diff --git a/Source/bmalloc/libpas/src/libpas/pas_utils_prefix.h b/Source/bmalloc/libpas/src/libpas/pas_utils_prefix.h
index c42ee5f1..fa450fe3 100644
--- a/Source/bmalloc/libpas/src/libpas/pas_utils_prefix.h
+++ b/Source/bmalloc/libpas/src/libpas/pas_utils_prefix.h
@@ -61,7 +61,13 @@ __PAS_BEGIN_EXTERN_C;
 #define PAS_BMALLOC_HIDDEN 1
 #endif
 
-#if defined(PAS_LIBMALLOC) && PAS_LIBMALLOC || defined(PAS_BMALLOC_HIDDEN) && PAS_BMALLOC_HIDDEN
+#if defined(__MINGW32__)
+/* PE/COFF has no visibility semantics and MinGW GCC emits
+   "visibility attribute not supported in this configuration; ignored"
+   for every single annotated declaration. libpas is linked as a static
+   archive, so there is nothing to decorate. */
+#define __PAS_API
+#elif defined(PAS_LIBMALLOC) && PAS_LIBMALLOC || defined(PAS_BMALLOC_HIDDEN) && PAS_BMALLOC_HIDDEN
 #define __PAS_API __attribute__((visibility("hidden")))
 #else
 #define __PAS_API __attribute__((visibility("default")))
-- 
2.47.3

